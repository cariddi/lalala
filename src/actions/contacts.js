
import { LIST_CONTACT_URL } from '../config/urls';
import { FETCH_CONTACTS } from './types';
	
export const fetchContacts = () => ( dispatch, _, api ) => dispatch( {
	type: FETCH_CONTACTS,
	promise: api
		.get( `${LIST_CONTACT_URL}` )
		.then( response => response.response )
	
} );