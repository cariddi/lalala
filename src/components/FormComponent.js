import React, { Fragment } from 'react'

function FormComponent({contactData}) {
    if(!contactData) return <h3>No contact data cat</h3>

    return (
        <Fragment>
            <div>Form Component</div>
            <h3>{contactData.name}</h3>
        </Fragment>
    )
}

export default FormComponent
