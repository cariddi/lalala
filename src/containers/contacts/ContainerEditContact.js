import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connectComponent } from '../../lib/connectComponent';
import { 
    getContactDataSelector,
    getCountriesSelector,
    getStatesSelector
} from '../../selectors/contacts';
// import FormComponent from '../../components/FormComponent';
import EditComponent from '../../components/EditComponent';

export class ContainerViewContact extends Component {
    componentDidMount() {
        const { id } = this.props.match.params;
		this.fetchContactData(id);
	}

    get contactData() { return this.props.contactData; }
    get countries() { return this.props.countries; }
    get states() { return this.props.states; }

    fetchContactData() { this.props.fetchContactData( this.props.match.params.id ); }
    fetchCountries() { this.props.fetchCountries(); }
    fetchStates() { this.props.fetchStates( this.contactData.country.id ); }

    editContactData() { this.props.editContactData( this.contactData.id, this.contactData ); }

    onFetchCountry = countryId => {
        this.props.fetchStates( countryId );
    }

    handleSubmit = formValues => {
        console.log('SUBMIT VALUES: ', formValues);
        this.editContactData( formValues.id, formValues );
    }

	render() {
        const { contactData, countries, states } = this;

		return (
            <EditComponent 
                initialValues={contactData} 
                onSubmit={this.handleSubmit} 
                onFetchCountry={this.onFetchCountry}
                countries={countries}
                states={states}
            />
        )
    }
}

ContainerViewContact.propTypes = {
    contactData: PropTypes.any,
    countries: PropTypes.any,
    states: PropTypes.any,
    
    fetchContactData: PropTypes.func.isRequired,
    fetchCountries: PropTypes.func.isRequired,
    fetchStates: PropTypes.func.isRequired,
    editContactData: PropTypes.func.isRequired
};

ContainerViewContact.defaultProps = {
    contactData: undefined,
    countries: undefined,
    states: undefined,
};

export default connectComponent( ( state, props ) => ( {
    fetchContactDataRequest: state.fetchContactDataRequest, //reducer
	fetchCountriesRequest: state.fetchCountriesRequest, //reducer
	fetchStatesRequest: state.fetchStatesRequest, //reducer
    editContactRequest: state.editContactRequest, //reducer
    contactData: getContactDataSelector( state, props ), //selector
    countries: getCountriesSelector( state, props ), //selector
    states: getStatesSelector( state, props ) //selector
} ) )( ContainerViewContact );
