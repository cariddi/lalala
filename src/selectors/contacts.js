import { createSelector } from 'reselect';

const getContacts = state => state.contacts;
const getContactData = ( state, props ) => {
    return props.match.params.id;
};

export const getContactsSelector = createSelector(
    getContacts,
    contacts => contacts.contacts
);

export const getContactDataSelector = createSelector(
    [ getContacts, getContactData ],
    ( contacts, contactID) => {
        if( !(contacts.contacts && contacts.contacts.length) ) return;
        
        return contacts.contacts.filter( c => c.id == contactID )[ 0 ];
    }
);

// let fakeState = {
//     contacts: [
//         {
//             "id": 5,
//             "name": "Susana Córdova Pabón",
//             "email": "ariadna_ruiz@abrego.es",
//             "phone1": "972853428",
//             "phone2": "631067997",
//             "birthdate": "1995-08-10",
//             "address": "Chalet Mónica Perea 7 Esc. 483",
//             "state": {
//                 "id": 29,
//                 "name": "Bahia"
//             },
//             "country": {
//                 "id": 3,
//                 "name": "Brasil"
//             },
//             "avatar": null
//         }
//     ]
// }

// console.log(getContactsKeysSelector(fakeState))