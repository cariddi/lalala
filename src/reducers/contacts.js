import { makeAsyncActionReducer, handleSuccess } from '../lib/reduxUtils';
import { FETCH_CONTACTS, FETCH_CONTACT_DATA, EDIT_CONTACT } from '../actions/types';

export const fetchContactsRequest = makeAsyncActionReducer( FETCH_CONTACTS );
export const fetchContactDataRequest = makeAsyncActionReducer( FETCH_CONTACT_DATA );
export const editContactRequest = makeAsyncActionReducer( EDIT_CONTACT );

export const contacts = ( state = {}, action ) => {
	switch ( action.type ) {
	case FETCH_CONTACTS:
		return handleSuccess( state, action, ( prevState ) => {
			const contacts = action.payload;
			return { ...prevState, contacts: contacts };
		} );
	case FETCH_CONTACT_DATA:
	case EDIT_CONTACT:
		return handleSuccess( state, action, ( prevState ) => {
			const contactData = action.payload;
			console.log('EDIT: ', contactData);
			if( !contactData ) return { ...prevState } ;
			if( !prevState.contacts ) prevState.contacts = [];
			if( !prevState.contacts.includes(contactData) ) prevState.contacts.push(contactData);
			
			return { ...prevState, contactID: contactData.id };
		} );
	default:
		return state;
	}
};
