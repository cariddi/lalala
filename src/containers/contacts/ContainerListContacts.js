import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connectComponent } from '../../lib/connectComponent';
import { getContacts } from '../../selectors/contacts';

export class ContainerListContacts extends Component {
    componentWillMount() {
		if ( !this.contacts ) { this.fetchContacts(); }
	}

    get contacts() { return this.props.contacts; }

    fetchContacts() { this.props.fetchContacts( ); }

    render() {
        const { contacts } = this;

        return (
            <div className='ListContacts'>
                Contact List
                {
                    this.props.contacts ?

                    this.props.contacts.map((c, idx) => {
                        return (
                            <li key={idx}>{c.id - c.name}</li>
                        )
                    })
                    :
                    <p>No contacts</p>
                }
            </div>
        )
    }
}

ContainerListContacts.propTypes = {
	contacts: PropTypes.any,
	fetchContacts: PropTypes.func.isRequired
};

ContainerListContacts.defaultProps = {
	contacts: undefined
};

export default connectComponent( ( state, props ) => ( {
	fetchContactsRequest: state.fetchContactsRequest, //reducer
	contacts: getContacts( state, props ) //selector
} ) )( ContainerListContacts );
