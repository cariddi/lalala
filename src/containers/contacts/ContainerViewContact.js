import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connectComponent } from '../../lib/connectComponent';
import { getContactDataSelector } from '../../selectors/contacts';
import FormComponent from '../../components/FormComponent'

export class ContainerViewContact extends Component {
    componentWillMount() {
        const { id } = this.props.match.params;
		// if ( !this.contactData ) { console.log('FETCH!!!'); this.fetchContactData(id); }
		this.fetchContactData(id);
	}

    get contactData() { return this.props.contactData; }

    fetchContactData() { this.props.fetchContactData( this.props.match.params.id ); }

	render() {
        const { contactData } = this;

		return (
            <div className='ListContacts'>
                <h2>Contact View</h2>
                <FormComponent contactData={contactData} />
            </div>
        )
    }
}

ContainerViewContact.propTypes = {
	contactData: PropTypes.any,
	fetchContactData: PropTypes.func.isRequired
};

ContainerViewContact.defaultProps = {
    contactData: undefined,
};

export default connectComponent( ( state, props ) => ( {
	fetchContactDataRequest: state.fetchContactDataRequest, //reducer
    contactData: getContactDataSelector( state, props ), //selector
} ) )( ContainerViewContact );
