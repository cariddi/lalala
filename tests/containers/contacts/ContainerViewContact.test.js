import React from 'react';
import { shallow } from 'enzyme';
import { ContainerViewContact } from '../../../src/containers/contacts/ContainerViewContact';
import FormComponent from '../../../src/components/FormComponent';

describe( 'ContactsView Container', () => {
    const fetchContactDataMocked = jest.fn();
	let container;

	beforeEach( () => {
		container = shallow(
			<ContainerViewContact 
                fetchContactData={fetchContactDataMocked}
                match={{ params: {id: 1} }}
			/>
		)
	} );

	it( 'passes the contact data to the Form component', () => {
		expect( container.find( FormComponent ) ).not.toBeNull();
		expect( container ).toMatchSnapshot();
	} );
} );
