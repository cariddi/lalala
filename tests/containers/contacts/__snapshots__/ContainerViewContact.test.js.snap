// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsView Container passes the contact data to the Form component 1`] = `
<div
  className="ListContacts"
>
  <h2>
    Contact View
  </h2>
  <FormComponent />
</div>
`;
